### PR TITLE
feat: support plugin_metadata of apisix

### PR DESCRIPTION
**Pull request by AlinsRan**
### Type of change:

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### What this PR does / why we need it:

* #1372

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

File: pkg/apisix/plugin_metadata.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+		log.Errorf("failed to create pluginMetadata: %s", err)
+		return nil, err
+	}
+
+	pluginMetadata, err := resp.Item.pluginMetadata()
+	if err != nil {
+		return nil, err
+	}
+	return pluginMetadata, nil
+}
```
Comment (tao12345666333) [inline pkg/apisix/plugin_metadata.go:135]: these should be removed.

**Review by tao12345666333 — CHANGES_REQUESTED** [review]

File: pkg/apisix/apisix.go [hunk 2]
```diff
@@ -152,6 +154,14 @@ type PluginConfig interface {
 	Update(context.Context, *v1.PluginConfig) (*v1.PluginConfig, error)
 }
 
+type PluginMetadata interface {
+	Get(context.Context, string) (*v1.PluginMetadata, error)
+	List(context.Context) ([]*v1.PluginMetadata, error)
+	Create(context.Context, *v1.PluginMetadata) (*v1.PluginMetadata, error)
+	Delete(context.Context, *v1.PluginMetadata) error
+	Update(context.Context, *v1.PluginMetadata) (*v1.PluginMetadata, error)
+}
```
[the remark this answers is not in the document]
Comment (AlinsRan) [reply pkg/apisix/apisix.go:163]: Yeah, but get has two interfaces
* plugin_name plugin_name: `/apisix/admin/plugin_metadata/{plugin_name}`
* All configurations: `/apisix/admin/plugin_metadata`

In fact, the http `PUT' method is used by create and update in all API (consumer/route...) interfaces.

[the remark this answers is not in the document]
Comment (AlinsRan) [reply pkg/apisix/apisix.go:163]: Here is the e2e verification of all interfaces.

[the remark this answers is not in the document]
Comment (tao12345666333) [reply pkg/apisix/apisix.go:163]: okay, we can keep the List

File: pkg/providers/k8s/configmap/configmap.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ctl.onAdd,
+			UpdateFunc: ctl.onUpdate,
+			DeleteFunc: ctl.onDelete,
+		},
+	)
+	return ctl
+}
+
+func (c *configmapController) Subscription(namespace, configName string) {
```
Comment (lingsamuel) [inline pkg/providers/k8s/configmap/configmap.go:73]: A function name should be a verb, I would like to use `Subscribe` here

File: pkg/providers/utils/manifest.go [hunk 4]
```diff
@@ -166,12 +166,38 @@ func DiffPluginConfigs(olds, news []*apisixv1.PluginConfig) (added, updated, del
...
+		oldMap[pm.Name] = pm
+	}
+	for _, pm := range news {
+		newMap[pm.Name] = pm
+	}
+
+	for _, pm := range news {
+		if ou, ok := oldMap[pm.Name]; !ok {
+			added = append(added, pm)
+		} else if !reflect.DeepEqual(ou, pm) {
```
Comment (lingsamuel) [inline pkg/providers/utils/manifest.go:182]: I think PluginMetadata is simple enough to compare values directly instead of using reflection.

**Review by lingsamuel — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]

**Pull request merged**